File: leetcodes/139. Word Break/wordBreak.java
class Solution {
  public boolean wordBreak(String s, List<String> wordDict) {
    //DP for an array of s.length() -> time O(mxn)
      //iterate over s and wordDict -> if word in wordDic len < s index number
        //if f[s index - word len] is true -> if (s substring(i - word len, i) == word) -> f[s index] = true, break
      //return dp[s.length()]
    boolean[] dp = new boolean[s.length() + 1];
    dp[0] = true;
    for(int i = 0; i <= s.length() ; i ++) {
      for (String word : wordDict){
        if (word.length() <= i) {
          if (dp[i - word.length()]){
            if (s.substring(i-word.length(), i).equals(word)){
              dp[i] = true;
              break;
            }
          }
        }
      }
    }
    return dp[s.length()];
  }
}

File: leetcodes/139. Word Break/wordBreak.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {boolean}
 */
 var wordBreak = function(s, wordDict) {
  //time O(nm) spaceO(n)
    let f = new Array(s.length + 1).fill(false)
    f[0] = true
    for (let i = 1; i <= s.length; i++) {
      for (let str of wordDict) { //based on the data type of wordDict,  if it is map, we can use contains
        if (str.length <= i) {
          if (f[i - str.length]) { //last str length position
            if (s.substring(i - str.length, i) === str) {
              f[i] = true //each length of str
              break;
            }
          }
        }
      }
    }
  return f[s.length]
};
